mysys_ssl/my_aes.cc

```diff
@@ -57,3 +57,97 @@ void my_aes_create_key(const unsigned char *key, uint key_length,
     *ptr^= *sptr;
   }
 }
+
+int my_aes_encrypt_nopad(const unsigned char *source, uint32 source_length,
```

Comment:
Can we implement this by using one of the AES block modes which do not require input message to be a multiple of AES_BLOCK_SIZE?
CFB, OFB or CTR for instance.

Reply:
`my_aes.h` only provides ECB and CBC for yassl. We build PS with OpenSSL, but I think we still need to be compatible. InnoDB page encryption using CBC. I would prefer CTR of course. MariaDB provides choice between CTR and CBC (CTR only available for file_key_management plugin though) with CBC being default. Trick with last partial block is taken from MariaDB.

Reply:
Guys, Mode of operations have nothing to do with block sizes. AES will always demand input to be multiple of 128 bits. If it is not multiple the padding is needed. Mode of operations (apart from ECB) are just the of getting random encryption. ECB is the only mode where no randomness is introduced into the output, thus each time when I encrypt text A with key K I will get B. ECB can be only used when text A is random - for instance INNODB is using it to encrypt innodb tablespace keys - and this is fine as keys are random. About which mode we should be using, 
there is a bit of discussion on this topic, but I think we should chose CBC. The reason being that in case of CTR we need to be sure to be always encrypting each input block with unique counter, the counter cannot repeat. In past gurus like Bruce Schneier (https://books.google.pl/books?id=nnqdQsrZnJgC&pg=PT141&lpg=PT141&dq=Bruce+Schneier+cbc+or+ctr&source=bl&ots=6GXUxllCq7&sig=ca_lQu9fkiu0Kf_qcBhhYkXVF2Q&hl=en&sa=X&ved=0ahUKEwiK-9Hy5c_XAhVKCpoKHXCZDOIQ6AEIXTAJ#v=onepage&q=Bruce%20Schneier%20cbc%20or%20ctr&f=false) would recommend CTR, but now they go with CBC. It safer this way with CTR it is quite possible one will repeat the counter.

Reply:
Streaming modes such as CTR or OFB do not require input to be multiple of block size. They actually perform AES operations on key and XOR plaintext with produced stream (somewhat simplified explanation :) ). So the resulting length is always the same as input length.

Reply:
Yeah, my bad. I forgot that also CTR and OFB can be used as stream ciphers.